Node sdk update (#1085)
rudderlabs/rudderstack-docs: docs/sources/event-streams/sdks/rudderstack-node-sdk.mdx

> description: Use RudderStack’s Node SDK to send events to various destinations.

Reviewer: ```suggestion
description: Use RudderStack’s Node SDK to send server-side events to various destinations.
```

Reviewer: To maintain consistency with the other server-side SDK metas.

## SDK setup requirements

1. Sign up to [RudderStack Cloud](https://app.rudderstack.com/signup).
2. <Link to="/dashboard-guides/sources/#adding-a-source">Set up a Node source</Link> in the dashboard. You should be able to see the <strong>write key</strong> for this source, as shown:

<img src="../../../assets/event-stream-sources/node-write-key-new.png" alt="Node source write key" />

You will also need the data plane URL. Refer to the <Link to="/dashboard-guides/overview/#data-plane-url">Dashboard Overview</Link> guide for more information on the data plane URL and where to find it.

<div class="successBlock">
The <strong>Setup</strong> tab in the RudderStack dashboard (seen above) has the SDK installation snippet containing both the write key and the data plane URL. Copy it to integrate the Node SDK into your application.
</div>

## Installing the Node.js SDK

> To install the RudderStack Node.js SDK using [npm](https://www.npmjs.com/), run the following command:

Reviewer: ```suggestion
To install the RudderStack Node SDK using [npm](https://www.npmjs.com/), run the following command:
```

Reviewer: For consistency.

const client = new Analytics(WRITE_KEY, DATA_PLANE_URL / v1 / batch)
```

</span>
      </TabPanel>
      <TabPanel>
<span>

```javascript
const Analytics = require('@rudderstack/rudder-sdk-node');

const client = new Analytics(WRITE_KEY, {
  dataPlaneUrl: DATA_PLANE_URL, // default: https://hosted.rudderlabs.com
});
```
</span>
      </TabPanel>
    </TabPanels>
</Tabs>

## SDK initialization options

> The RudderStack Node.js SDK provides the following parameters which can be passed during the SDK initialization:

Reviewer: ```suggestion
The RudderStack Node SDK provides the following parameters which can be passed during the SDK initialization:
```

| Parameter                   | Data type     | Default value    | Description         |
| :--------------------- | :------- | :------------------------------ | :------------------ |

> | `flushAt`              | Integer   | 20                              | Number of events to be flushed when reached this limit.   |

Reviewer: A bit confusing - do we mean the SDK flushes when this number of events is elapsed?

Reviewer 2: Yes

| `flushInterval`        | Integer   | 10000                           |  Maximum timespan (in milliseconds) after which the events from the in-memory queue are flushed. |
| `maxInternalQueueSize` | Integer   | 20000                           |  Maximum length of the in-memory queue.     |

> | `logLevel`             | String   | 'info'                          | Log level. For example: `debug`, `error`, etc.     |

Reviewer: ```suggestion
| `logLevel`             | String   | 'info'                          | Sets the logging level. The acceptable values are `info`, `debug`, `error`, etc.     |
```

Reviewer: @ishachopra0401 @MoumitaM LMK if this looks ok - if yes, can we add the rest of the acceptable values for this param?

Reviewer 2: LGTM

> The following initialization parameters are only available for RudderStack Node.js SDK **v2 and above**:

Reviewer 2: ```suggestion
The following initialization parameters are only available for RudderStack Node.js SDK **v2.x.x and above**:
```

> The following initialization parameters are only available for RudderStack Node.js SDK **v2 and above**:

Reviewer: ```suggestion
The following initialization parameters are only available for RudderStack Node SDK **v2 and above**:
```

| Parameter                   | Data type     | Default value    | Description         |
| :--------------------- | :------- | :------------------------------ | :------------------ |
| `dataPlaneUrl`         | String   | `https://hosted.rudderlabs.com` | Data plane URL.        |

> | `path`                 | String   | `/v1/batch`                     | Path to batch endpoint.          |

Reviewer: ```suggestion
| `path`                 | String   | `/v1/batch`                     | Path to the `batch` endpoint.          |
```

> | `maxQueueSize`         | Integer   | 460800 (500KB)                   | Maximum payload size of a batch request.   |

Reviewer: ```suggestion
| `maxQueueSize`         | Integer   | 460800 (500KB)                   | Maximum payload size of a `batch` request.   |
```

| `axiosConfig`          | Object   | N/A                             | Axios configuration.            |
| `axiosInstance`        | Object   | N/A                             | Axios instance.              |
| `axiosRetryConfig`     | Object   | N/A                             | Axios retry configuration.          |

> | `retryCount`           | Integer   | 3                               | Number of times a request is retried by axios if failed.  |

Reviewer: ```suggestion
| `retryCount`           | Integer   | 3                               | Number of times a request is retried by Axios in case of failure.  |
```

</div>

If the Node.js SDK fails to deliver the event data to RudderStack in first attempt, it retries the event delivery. However, if RudderStack is unavailable for a longer duration, there is a possibility of data loss. To prevent this scenario, RudderStack provides the data persistence feature where the event data is persisted in **Redis**, guaranteeing event delivery. Simultaneously, the SDK can retry multiple times as the queue is maintained in a different process space (in this case, Redis).

<div class="infoBlock">
To use this feature, you will need to host a Redis server to use it as the intermediary data storage queue. RudderStack uses <a href="https://github.com/OptimalBits/bull">Bull</a> as the interface layer between the Node.js SDK and Redis.
</div>

To achieve data persistence, you need to call the `createPersistenceQueue` method which takes two parameters - `queueOpts` and `callback`. It initializes the persistent queue. A sample SDK initialization is shown below:

<Tabs>
  <TabList>
    <Tab>Below v2.0.0</Tab>
    <Tab>v2.0.0 and above</Tab>
  </TabList>
    <TabPanels>
      <TabPanel>
<span>

```jsx

> const client = new Analytics("write_key","server_url/v1/batch",{

Reviewer 2: ```suggestion
const client = new Analytics("write_key","DATA_PLANE_URL/v1/batch",{
```

The `alias` method parameters are as mentioned below:

| **Field**       | **Type** |  **Description**           |
| :------------ | :------- | :------------------- |

> | `userId` <br/> <span style="color: #4D4DFF;font-size:12px;">Required, if <strong>anonymousId</strong> is absent.</span>       | String | Unique identifier for a user in your database.                                            |

Reviewer 2: ```suggestion
| `userId` <br/> <span style="color: #4D4DFF;font-size:12px;">Required</span>       | String | Unique identifier for a user in your database.                                            |
```

> | `anonymousId` <br/> <span style="color: #4D4DFF;font-size:12px;">Required, if <strong>userId</strong> is absent.</span>   | String | The SDK automatically sets this identifier in cases where there is no unique identifier for the user.|

Reviewer 2: ```suggestion
| `anonymousId` <br/> <span style="color: #4D4DFF;font-size:12px;">Optional</span>   | String | The SDK automatically sets this identifier in cases where there is no unique identifier for the user.|
```

| `previousId` <br/> <span style="color: #4D4DFF;font-size:12px;">Required</span>  | String   | The previous unique identifier of the user.       |
| `context`      | Object   | An optional dictionary of information that provides context about a message. It is not directly related to the API call.  |
| `integrations` | Object   |  An optional dictionary containing the destinations to be either enabled or disabled.   |

> | `traits`       | Object   | An optional dictionary of the group's traits like `name`or `email`.     |

Reviewer: ```suggestion
| `traits`       | Object   | An optional dictionary of the user's traits like `name`or `email`.     |
```

| `timestamp`    | Date     | The timestamp of the message's arrival.   |

> ## Node.js SDK data persistence

Reviewer: ```suggestion
## Data persistence
```

Reviewer: Thoughts?

Reviewer 2: LGTM

<div class="warningBlock">
This is a beta feature. Contact the RudderStack team on <a href="https://rudderstack.com/join-rudderstack-slack-community">Slack</a> if you face any issues.
</div>

> If the Node.js SDK fails to deliver the event data to RudderStack in first attempt, it retries the event delivery. However, if RudderStack is unavailable for a longer duration, there is a possibility of data loss. To prevent this scenario, RudderStack provides the data persistence feature where the event data is persisted in **Redis**, guaranteeing event delivery. Simultaneously, the SDK can retry multiple times as the queue is maintained in a different process space (in this case, Redis).

Reviewer: ```suggestion
If the Node SDK fails to deliver the events to RudderStack in the first attempt, it retries the event delivery. However, if RudderStack is unavailable for a longer duration, there is a possibility of data loss. To prevent this scenario, the Node SDK has the data persistence feature where the event data is persisted in **Redis**, guaranteeing event delivery. Simultaneously, the SDK can retry multiple times as the queue is maintained in a different process space (Redis).
```

<div class="infoBlock">

> To use this feature, you will need to host a Redis server to use it as the intermediary data storage queue. RudderStack uses <a href="https://github.com/OptimalBits/bull">Bull</a> as the interface layer between the Node.js SDK and Redis.

Reviewer: ```suggestion
To use this feature, you will need to host a Redis server to use it as the intermediary data storage queue. RudderStack uses <a href="https://github.com/OptimalBits/bull">Bull</a> as the interface layer between the Node SDK and Redis.
```
